Commit version bump bug fix and enhancements

jdee opened the pull request:
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] -->

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

This is motivated by Issue #9341. The problem is that the `:settings` option to `commit_version_bump` doesn't work if the Settings.bundle is in a group in the Xcode project. This option has been there for some time, but it was broken when I first tried to use it. I "fixed" that bug at some point, resulting in the faulty code that's there now. This originally showed up as a bug in my settings_bundle plugin. That has been fixed but not yet released, pending resolution of the issue with `commit_version_bump`.

(_Edited_) To produce the bug:

```
git clone https://github.com/jdee/settings-bundle
cd settings-bundle
git checkout commit-version-bump-bug
bundle install
bundle exec fastlane group_test
```

To test the fix:

```
git checkout -- .
git checkout commit-version-bump-bug-fixed
bundle install
bundle exec fastlane group_test
```

Note that `bundle exec fastlane test` will succeed in both branches. The difference between SettingsBundleExample and SettingsBundleGroupExample is that in the former, the Settings.bundle is at the root of the project in Xcode, but in the latter it's under the SettingsBundleGroupExample group. The `#path` method used [here](https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/commit_version_bump.rb#L220) returns a value relative to the group, not the project root.

But a bigger issue is that this logic belongs in the plugin, not in this action. I suggested two options in the issue. I've implemented both. There are three stages to the fix, represented by different commits.

1. Fix the broken logic in `settings_bundle_file_path`.
2. Add an `:include` option to specify arbitrary files to include.
3. Add `lane_context[SharedValues::MODIFIED_FILES]` and `Actions::add_modified_files`.

If only some of these changes should be merged, they can easily be reverted or removed.

### Description

1. Fix the broken logic in `settings_bundle_file_path`:
  Use the `#real_path` method, which returns the absolute path to the file. This fixes the original
  bug with a tiny change.
2. Add an `:include` option to specify arbitrary files to include.
  This allows explicit inclusion of an array of relative file paths. This has the virtue that there is no
  mystery if a file is included, but more parameters are required, and the correct path has to be
  explicitly supplied. There is already an `:ignore` parameter. This parameter seems to complement
  it.
3. Add `lane_context[SharedValues::MODIFIED_FILES]` and `Actions::add_modified_files`.
  This allows any action that modifies files to be included in a commit to pass them directly to
  the `commit_version_bump` action via the lane_context. It allows the context to accumulate
  the unique results of all actions. It eliminates the need for any additional parameters to support
  plugins that make modifications to include in a version bump. It could easily be used to support
  Android as well, rewrite `commit_version_bump` and obviate the `:xcodeproj` and `:settings`
  parameters.

I mentioned in the issue that it's possible for a settings bundle to have a custom name other than Settings.bundle. This is and isn't true. It's possible to create such a thing, but it doesn't show up in the Settings app unless the name is Settings.bundle. So that's not a concern, and the first commit is enough to fix the issue and probably put the question of settings bundles to rest.

But it seems plausible that there may be other cases where a lane might modify files other than the Xcode project and Info.plist as part of a version bump. Or I might have to fix another bug in my plugin and have to change `commit_version_bump` again. A more extensible approach seems like a good idea. The other two options are presented for consideration.

(_Edited_) To try out these options, use the `commit-version-bump-bug-fixed` branch in the plugin repo above. See the [Fastfile](https://github.com/jdee/settings-bundle/blob/commit-version-bump-bug-fixed/fastlane/Fastfile) there, which has three lanes that demonstrate the three different methods.

Using the `:include` option:
```
bundle exec fastlane test
```

Using the corrected `:settings` option:
```
bundle exec fastlane group_test
```

Using the lane_context (no `:include` or `:settings` option):
```
bundle exec fastlane library_test
```

Note: `git reset --hard HEAD^` erases the last commit in the current branch and is useful for testing this action.

jdee committed:
#9341: Corrected inference of Settings.bundle path using #real_path method

jdee committed:
Added :include option to commit_version_bump

jdee committed:
Updated failing test to supply #real_path

jdee committed:
Added test for :include parameter

jdee committed:
Added SharedValues::MODIFIED_FILES and Actions::add_modified_files

jdee committed:
rubocop corrections

jdee committed:
Added code examples

jdee commented:
@googlebot I signed it.

jdee commented:
There are now two branches in the settings-bundle repo: `commit-version-bump-bug` and `commit-version-bump-bug-fixed`. The new options are used in the second branch. The original instructions above have been updated.

ohayon commented on fastlane/lib/fastlane/actions/commit_version_bump.rb:
```diff
@@ -201,23 +234,46 @@ def self.category
 
-          project_parent = File.dirname project.path
-          File.join(project_parent, settings_bundle.path, settings_file_name)
+          File.join settings_bundle.real_path, settings_file_name
```
Could you put parentheses in here?

ohayon reviewed (changes requested):
Hey @jdee, thank you so much for this PR! I would really like to merge just the first commit/change that you made and then address the second and third in other PRs if that is possible. Would you mind opening up other PRs as those kind of change/add some functionality where it appears that the first really just fixes the bug? Let us know if you have any questions! 🚀

jdee committed:
Add examples of current :settings option

jdee committed:
Revert "Added code examples"

This reverts commit 223171406cfa376c523faf0c5b84a1129199e865.

jdee committed:
Revert "rubocop corrections"

This reverts commit 4572f9da9607bdf5d447edac66df02d53a8b2e12.

jdee committed:
Revert "Added SharedValues::MODIFIED_FILES and Actions::add_modified_files"

This reverts commit e16372e7b8f03d37cdffce7e1f4531426bfaf5c8.

jdee committed:
Revert "Added test for :include parameter"

This reverts commit bc8ce5f24fd12b53fdf9761670fb9e05134f1f19.

jdee committed:
Revert "Added :include option to commit_version_bump"

This reverts commit 35c72c1189971363a595e6c18471fd9b28646456.

jdee committed:
Add parentheses to File.join call

jdee commented:
@ohayon: Done. The first commit broke a unit test, fixed by the third. I also fixed the parentheses and added examples of the current usage of the `:settings` option, which I neglected to do the last time around. Would you prefer one PR or two for the enhancements?

ohayon reviewed (approved):
Thanks for making those changes @jdee !! 🚀

ohayon commented:
@jdee - I think one PR for the enhancements would be great! Thanks so much for being flexible!
